### PR TITLE
feat: `WalReplicator` skeleton implementation

### PR DESCRIPTION
**Pull request by waynexia**
Signed-off-by: Ruihang Xia <waynestxia@gmail.com>

# Which issue does this PR close?

ref #175

# Rationale for this change
 

Follow the context from #175. `ReaderTable` needs to get the newest data from the corresponding `WriterTable` via WAL. This implements a replicator that keeps polling data from WAL.

This replicator works as following:
```
          register IDs
         need replicate
     ┌─────────────────────┐
     │                     │
     │              ┌──────▼───────┐
┌────┴─────┐        │  background  │
│Role Table│        │WAL Replicator│
└────▲─────┘        └──────┬───────┘
     │                     │
     └─────────────────────┘
          replicate log
            to table
```
A background replicate task is starting with `Instance`. `Instance` can register table to it with `register_table()`. The replicator will loop on the registered table to poll logs from WAL. Invalid table (those tables with incorrect states) will be removed from the queue.

There are two problems I can think of.
- A table may be kicked out in accident without notification.
- The entire replicate procedure inside a replicator is serialized. 

# What changes are included in this PR?

Implement the replicator described above.

# Are there any user-facing changes?

No

# How does this change test

New added unit test. The add/evict table logics need a `RoleTable` trait definition, I'm going to cover this part after we got one.

### CODE REVIEW

File: analytic_engine/src/wal_replicator.rs [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+                .await
+                .is_ok()
+            {
+                info!("WAL Replicator stopped");
+                break;
+            }
+        }
+    }
+
+    async fn consume_iter(
```
Comment (ShiKaiWi) [inline analytic_engine/src/wal_replicator.rs:230]: Is it better if we call this method `consume_logs`?

File: analytic_engine/src/wal_replicator.rs [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+
+        // update sequence number in state
+        replicate_state
+            .last_synced_seq
+            .fetch_max(max_seq, Ordering::Relaxed);
+
+        Ok(())
+    }
+
+    async fn replay_log(
```
Comment (ShiKaiWi) [inline analytic_engine/src/wal_replicator.rs:267]: `replay_logs` seems better.

File: analytic_engine/src/wal_replicator.rs [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+        &self,
+        log: &mut VecDeque<LogEntry<ReadPayload>>,
```
Comment (ShiKaiWi) [inline analytic_engine/src/wal_replicator.rs:269]:
```suggestion
        logs: &mut VecDeque<LogEntry<ReadPayload>>,
```

File: analytic_engine/src/wal_replicator.rs [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+        let join_handle = runtime.spawn(
+            self.inner
+                .clone()
+                .start_replicate(self.stop_receiver.take().unwrap()),
+        );
+        *self.join_handle.lock().await = Some(join_handle);
+    }
+}
+
+pub struct WalReplicatorInner {
```
Comment (ShiKaiWi) [inline analytic_engine/src/wal_replicator.rs:149]: Shall we expose this struct? Maybe we call it `Inner` if it is private.

**Review by ShiKaiWi — APPROVED** [review]

File: analytic_engine/src/wal_replicator.rs [hunk 5]
```diff
@@ -0,0 +1,380 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+//! WAL Replicator implementation.
```
Comment (chunshao90) [inline analytic_engine/src/wal_replicator.rs:3]: Rename `Replicator` to `Consumer`.

File: analytic_engine/src/wal_replicator.rs [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+
+                // double check state before writing to table. Error due to
+                // state changed after this check will be treat as normal error.
+                if !state.check_state() {
+                    invalid_regions.push(state.region_id);
+                    continue;
+                }
+
+                // read logs from iterator
+                if let Err(e) = self.consume_logs(&mut iter, state).await {
```
[the remark this answers is not in the document]
Comment (waynexia) [reply analytic_engine/src/wal_replicator.rs:212]: https://github.com/CeresDB/ceresdb/pull/179#discussion_r943123353

tired of changing naming again and again😣  leave it be if one name is style compatible and isn't ambiguous

File: analytic_engine/src/wal_replicator.rs (hunk 5 above)
Comment (waynexia) [reply analytic_engine/src/wal_replicator.rs:3]: Could you please describe why `Consumer` is better than `Replicator`? My reason for using `Replicator` is stated in #175. TL;DR: A replicator replicates logs from WAL to tables, it keeps the data between `Leader` table and its replicas in sync.

File: analytic_engine/src/wal_replicator.rs [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+        let mut max_seq = SequenceNumber::MIN;
+
+        while should_continue {
+            // fetch entries
+            buf = iter
+                .next_log_entries(WalDecoder::default(), buf)
+                .await
+                .context(ReadWal)?;
+            if buf.len() <= self.config.batch_size {
+                should_continue = false;
```
[the remark this answers is not in the document]
Comment (waynexia) [reply analytic_engine/src/wal_replicator.rs:246]: `buf` is and needs to be consumed at the end of each loop. I need a flag to record whether the loop can keep going based on the length of `buf` I polled. `should_continue` is that flag.

File: analytic_engine/src/wal_replicator.rs (hunk 5 above)
Comment (chunshao90) [reply analytic_engine/src/wal_replicator.rs:3]: `WAL` is divided into `log replicator` and `log consumer`. Currently, the log replicator is implemented by wal mod such as `WAL based on OBKV`, and followers only need to consume logs. In `start_replicate` func, actually calls `consume_log` not `replicate_log`.

Comment (waynexia) [reply analytic_engine/src/wal_replicator.rs:3]: Or I can interpret `Consumer` as the one that eats logs and generate new data, which is `Table` in `Instance`. Your explanation might be right but I don't think it's better. And if so then we can keep it.

Furthermore in our model, distributed WAL component like `OBKV` is a place to store the write ahead logs. It doesn't "replicate" logs to other nodes neither by definition nor in practice. It's those followers that actually copy logs from storage to itself. IMHO calling this copy procedure "replicate" has no problem.

Comment (chunshao90) [reply analytic_engine/src/wal_replicator.rs:3]: @ShiKaiWi  what you think?

Comment (ShiKaiWi) [reply analytic_engine/src/wal_replicator.rs:3]: Here are my some thoughts about this naming issue:
- `Consumer` is actually more suitable for a more common case (only read logs without knowing how to apply these log contents to table).
- `Replicator` basically denotes makes a redundant copy of the data without changes about the original data, but here the data has been changed into `Table`'s data from the log entries.

So I propose a new name `Synchronizer`: synchronize logs from WAL to table.

Comment (waynexia) [reply analytic_engine/src/wal_replicator.rs:3]: `Synchronizer` looks good to me. How about @chunshao90

**Review by ShiKaiWi — APPROVED** [review]
LGTM

**Pull request merged**